### ISSUE TITLE
Add "full" transparent mode that also spoofs the source address

### ISSUE THREAD

**Issue by yoavfrancis** [mitmproxy/mitmproxy#436]
Hi there,

I was wondering - I'm running mitmproxy on a Linux machine with 2 NIC's (1 goes to clients and the other goes to the internet) - is there an option for "full transparency" - That is, that packets from the mitm to the internet will be sent with the client's src IP (and even MAC)? 

If not - I'd be more than happy to hear tips on how to implement this.

(Also, out of curiosity - What would happen if the Internet NIC on the Linux machine is not given an IP? could the mitm-to-internet connection still work than?)

Thanks!
Yoav

**Comment by mhils** [comment]
Hi Yoav,

sorry for the late reply - you are not the first one to ask this and I asked for permission to share the following mail first :wink: 

Paul Pearce wrote:
> Hey folks,
>
>Have you guys given any thought to adding support for full transparent
proxying? What mitmproxy supports now is sometimes referred to as
"client transparent" proxying. "Full transparent" proxying generally
means the server can't identify that a proxy exists. Practically
speaking, it means spoofing the client IP and doing some routing magic
to get the packets back to mitmproxy.
>
>Some other software that supports this:
http://thewalter.net/stef/software/proxsmtp/transparent-full.html
http://blog.loadbalancer.org/configure-haproxy-with-tproxy-kernel-for-full-transparent-proxy/
>
>From a quick glance, I think changes would need to be made from
mitmproxy -> libmproxy (proxy.py's invokation of netlib's
tcp.TCPServer) -> netlib.tcp.TCPServer.connect(). With an add of:
>
>connection.setsockopt(socket.SOL_IP, 19, 1)
>
>(the transparent constant isn't in python < 3.3, IIRC. RE the
cosntant, see https://www.kernel.org/doc/Documentation/networking/tproxy.txt)
>
>Does this sound like something you'd like to support?
>
>I threw together a proof of concept patch.
> https://gist.github.com/mhils/d37ec10401eed8f8c797

Maximilian Hils wrote:
> Generally speaking, we're always happy to accept pull requests, but I'm a bit afraid that this may confuse a lot of users ("of course I want *FULL* transparent proxying!") while providing little value to most. We already have many bug reports that result from a misunderstanding of what transparent mode actually does. 
> I think calling it '--spoof-client-ip' would be a viable alternative. Would you be willing to submit a PR for this @ GitHub?

TL;DR: Yes it is possible and we'd love to have a PR for this. :-)

Cheers,
Max

**Comment by aleonmur** [comment]
Hi,

I'm also interested in this feature, did you manage to make it work?

If not, I would appreciate any pointers you could give me.

Cheers,
Alvaro
